### PR TITLE
RPG and minor improvements.

### LINKED ISSUE #226: [💡SUG] RPG: Generating Long Semantic IDs in Parallel for Recommendation, KDD '25
**Describe the model you wish to add and its strengths.**
RPG, a lightweight framework for semantic ID-based recommendation. The key idea is to produce unordered, long semantic IDs, allowing the model to predict all tokens in parallel. 

**Describe the type of model**
Please specify whether the model is:

    Context-aware

    Knowledge-aware

    Path-reasoning

    General-purpose

    Other (please specify): Sequential Recommender

**Describe the loss of model**
MTP Loss. Basically, it computes the loss as the sum of the CE for each prediction head.

**Describe the Input of model**
sequence of items

**Provide the original paper or source**
Generating Long Semantic IDs in Parallel for Recommendation (https://arxiv.org/abs/2506.05781), KDD '25

**Code availability**
https://github.com/facebookresearch/RPG_KDD2025

**Python and PyTorch compatibility**
NA

**Additional context**
NA

### PR DESCRIPTION
**Pull request by alessandrosocc**
- Added RPG (https://arxiv.org/abs/2506.05781, https://github.com/facebookresearch/RPG_KDD2025)
- Added tests for RPG
- Added support for using sequential datasets without RecBole's data augmentation
- Print a table to better visualize the results when len(topk)>1
- fixed benchmark_filename usage in sequential recommender
- when benchmark_filename has size 2 (meaning train and test split), the validation results are calculated on the test set (not the best, but I assume that nobody will do hyperparameters tuning on that set).
- added benchmark_item_filename to specify the subset of items that can be recommended (filtered) during test time.
- added the fix to kg_dataset.py filter_link.py from #243 by @jackmedda 
- fixed benchmark_filename_list behavior in sequential recommenders
- added an argument "AUGMENT_ITEM_SEQ" that regulates whether to use the standard sequence creation process of recbole or not (FIRST DRAFT, currently not completed)
- added the possibility to hide the configuration output through the "show_configuration" argument
- added the possibility to show the output of the training through the "verbose" argument
- added the similarity model that provide user recommendation through embedding similarity
- in addition to showing the total dataset statistics, we also show statistics for each data split
- added support for custom data loaders
- minor fixes... 🤖

### CODE REVIEW

File: hopwise/trainer/trainer.py [hunk 1]
```diff
@@ -246,7 +229,7 @@ def _train_epoch(self, train_data, epoch_idx, loss_func=None, show_progress=Fals
                 self.set_reduce_hook()
                 sync_loss = self.sync_grad_loss()
 
-            with autocast(device_type=self.device.type, enabled=self.enable_amp):
+            with torch.autocast(device_type=self.device.type, enabled=self.enable_amp):
```
[the remark this answers is not in the document]
Comment (alessandrosocc) [reply hopwise/trainer/trainer.py:232]: check the warning box here: https://docs.pytorch.org/docs/stable/amp.html

File: hopwise/model/sequential_recommender/rpg.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            padding = torch.full((item_seq.size(0), 1), -torch.inf, device=item_seq.device)
+            scores = torch.cat([padding, scores], dim=1)
+
+        return scores
+
+    def graph_propagation(self, token_logits):
+        """
+        I don't like this implementation. It propagates only on a random subset
+        (#self.n_beams) of nodes. Does it make sense?
+        """
```
Comment (jackmedda) [inline hopwise/model/sequential_recommender/rpg.py:176]: It seems it behaves similarly to a top-p sampling. First, it takes from the graph only the topk indices (such as taking the probability p higher than a certain threshold), then it randomly samples from these filtered indices represented by the `adjacency` object. There may be alternative solutions, similar to HF beam search that using probabilities constrained on previous choices.

Nonetheless, this is a personal comment and shouldn't stay here, let alone as a docstring.
Maybe reformulate as `# TODO: is the propagation on a random subset sound?`

File: hopwise/properties/model/BPR.yaml [hunk 3]
```diff
@@ -1 +1,2 @@
-embedding_size: 64              # (int) The embedding size of users and items.
\ No newline at end of file
+embedding_size: 64              # (int) The embedding size of users and items.
+epochs: 1
```
Comment (jackmedda) [inline hopwise/properties/model/BPR.yaml:2]: Forgot to delete it?

File: hopwise/properties/model/PGPR.yaml [hunk 4]
```diff
@@ -3,11 +3,14 @@ state_history: 1
 max_acts: 250
 gamma: 0.99
 action_dropout: 0.0
-hidden_sizes: [512,256]
+hidden_sizes: [256,128]
 max_path_len: 3
 weight_factor: 1e-3
 beam_search_hop: [25,5,1]
 fix_scores_sorting_bug: False
+learning_rate: 1e-4
```
Comment (jackmedda) [inline hopwise/properties/model/PGPR.yaml:11]: learning_rate isn't specific to PGPR. Better always to experiment with an external config file or passing parameters through CLI during development

File: hopwise/properties/model/PGPR.yaml [hunk 5]
```diff
@@ -3,11 +3,14 @@ state_history: 1
 max_acts: 250
 gamma: 0.99
 action_dropout: 0.0
-hidden_sizes: [512,256]
+hidden_sizes: [256,128]
```
Comment (jackmedda) [inline hopwise/properties/model/PGPR.yaml:6]: Does this modification align with the original paper?

File: hopwise/properties/model/RPG.yaml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+#   decoding graph construction
+chunk_size: 1024
+num_workers: 64
+
+eval_args:
+  split: {'LS': 'valid_and_test'}
+  order: TO
+  mode: full
+train_neg_sample_args: ~
+reproducible: True
```
Comment (jackmedda) [inline hopwise/properties/model/RPG.yaml:36]: What is the attribute `reproducible` used for? I remember Recbole has one named `reproducibility`, but I don't remember this.

File: hopwise/properties/model/RPG.yaml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+MAX_ITEM_LIST_LENGTH: 50
+AUGMENT_ITEM_SEQ: False
+
+# Semantic ID configs (FAISS OPQ)
+n_codebook: 4
+codebook_size: 256
+faiss_omp_num_threads: 32
+
+# RPG configs
+use_gcd: False # use graph constrained decoding
```
Comment (jackmedda) [inline hopwise/properties/model/RPG.yaml:21]: Isn't the GCD a default component of RPG? Asking for curiosity

File: hopwise/trainer/trainer.py [hunk 8]
```diff
@@ -1864,7 +1859,7 @@ def _train_epoch(self, train_data, epoch_idx, loss_func=None, show_progress=Fals
                 train_data,
                 total=len(train_data),
                 ncols=100,
-                desc=set_color(f"Train {epoch_idx:>5}", "magenta", progress=True),
+                desc=set_color(f"Train {epoch_idx:>5}", "w", progress=True),
```
Comment (jackmedda) [inline hopwise/trainer/trainer.py:1862]: Typo?

File: hopwise/utils/logger.py [hunk 9]
```diff
@@ -45,8 +43,9 @@ def __call__(self, *args, **kwargs):
 
 
 def progress_bar(*args, **kwargs):
+    global _progress_bar  # noqa: PLW0603
     if _progress_bar is None:
-        return ProgressBar({"progress_bar_rich": True})(*args, **kwargs)
+        _progress_bar = ProgressBar(progress_bar_rich=not os.environ.get("DISABLE_RICH", False))
```
Comment (jackmedda) [inline hopwise/utils/logger.py:48]: This seems fine, but `global` should be used to specify that you mean to use the global variable, which must be defined outside, so `_progress_bar = None` shouldn't have been removed.
Line 48 needs `# noqa: PLW1508` as suggested by GitHub, due to a weird behavior where the linter expects `False` to be a string, but doing so doesn't actually modify the env variable.

File: tests/model/test_model_auto.py [hunk 10]
```diff
@@ -812,6 +828,10 @@ def test_s3rec(self):
             }
             quick_test(config_dict)
 
+    def test_rpg(self):
+        config_dict = {"model": "RPG", "train_neg_sample_args": None, "n_codebook": 2, "codebook_size": 32}
+        quick_test(config_dict)
+
```
Comment (jackmedda) [inline tests/model/test_model_auto.py:834]: I guees we need more tests for RPG. For instance, one `_with_GCD` and maybe other ones if some components can be toggled.

**Review by jackmedda — CHANGES_REQUESTED** [review]

File: hopwise/data/dataset/sequential_dataset.py [hunk 11]
```diff
@@ -138,15 +134,6 @@ def data_augmentation(self):
...
 
-    def _benchmark_presets(self):
-        list_suffix = self.config["LIST_SUFFIX"]
-        for field in self.inter_feat:
-            if field + list_suffix in self.inter_feat:
-                list_field = field + list_suffix
-                setattr(self, f"{field}_list_field", list_field)
-        self.set_field_property(self.item_list_length_field, FeatureType.TOKEN, FeatureSource.INTERACTION, 1)
-        self.inter_feat[self.item_list_length_field] = self.inter_feat[self.item_id_list_field].transform(len)
-
```
[the remark this answers is not in the document]
Comment (alessandrosocc) [reply hopwise/data/dataset/sequential_dataset.py:149]: yes, self.item_id_list_field is never created when using benchmark_presets, so It raise an error. The whole setting for benchmark files on sequential recommender doesn't work.
